Extend env vars and user expansion for work_dir and cry_copy_dir

pkuderov opened the pull request:
This PR extends the expansion logic for environment variables (`$XXX`) and user home directories (`~`) to include the `work_dir` and `cry_copy_dir` fields, ensuring these paths are fully resolved and interpreted correctly. Readme is updated accordingly to mention it, with examples.

Additional improvements:
- **Pylint fixes**: Addresses and resolves all outstanding pylint warnings for cleaner, compliant code.
- **Improved logging for unexpanded variables**: Adds explicit warnings when expanded strings still contain a `$`, which may indicate missing or misconfigured environment variables (e.g. due to a failed or forgotten rc-file sourcing). This helps prevent silent misconfigurations.
- **Unified path formatting and sanitization**: Standardizes how path-like variables are processed and normalized, reducing the likelihood of bugs caused by redundant or missing normalization and resolution steps.
- **Bug fix for work_dir resolution**: Corrects an issue where `work_dir` could incorrectly resolve to its parent directory in certain cases.

pkuderov committed:
support expansion on tuple, list, dict and simple types; apply expansion to environment, work_dir, cry_copy_dir, exclude_from_copy

pkuderov committed:
add work_dir, cry_copy_dir path auto-sanitizing before processing

pkuderov committed:
remove redundand path resolving, switch to using unified resolved str paths

pkuderov committed:
remove exclude_from_copy from var/user expansion as not useful

pkuderov committed:
warn user on non-expanded '$' signs, add test for warning, polish code, fix pylint complaints

pkuderov committed:
update and polish readme regarding env vars/user expansion

pkuderov commented:
Hey folks, I can try adding some tests as well — but it’ll take a bit of time to figure out the right way to mock file-related stuff. I don’t quite remember off the top of my head.

pkuderov committed:
reduce the number of lines for the non-expanded vars warning msg

mrsndmn commented on cryri/utils.py:
```diff
@@ -1,64 +1,131 @@
+    cfg.cry_copy_dir = expand_vars_and_user(cfg.cry_copy_dir)
+
+
+def sanitize_config_paths(cfg: ContainerConfig):
```
nitpick: Would you like to move `sanitize_config_paths` and `expand_config_vars_and_user` to `class ContainerConfig` as object methods?

pkuderov replied:
sounds good! But there will be a small problem — a circular imports. So, either local [func] imports in these methods, or moving these funcs to another `config_utils.py` or to `config.py` itself. What do you think?

pkuderov replied:
sounds good! But there will be a small problem — circular imports. So, either local [func] imports in these methods, or moving these funcs to another `config_utils.py` or to `config.py` itself. What do you think?

pkuderov replied:
sounds good! But there will be a small problem — a circular imports. So, either local [func] imports in these methods, or moving these funcs to another `config_utils.py` or to `config.py` itself. What do you think?

pkuderov replied:
sounds good! But there will be a small problem — circular imports. So, either local [func] imports in these methods, or moving these funcs to another `config_utils.py` or to `config.py` itself. What do you think?

mrsndmn replied:
Good catch!

I believe that utils.py shouldn't depend on `config.py`.

**Proposal**:
* Move all these functions to `config.py`, including `create_job_description`.
* Keep only create_run_copy in `utils.py`, but modify its interface to remove the `CryConfig` dependency:

```python
def create_run_copy(work_dir, cry_copy_dir, exclude_from_copy=None) -> str:
    """Create a copy of the work directory for the run."""
    now = datetime.now()
    now_str = now.strftime(DATETIME_FORMAT)
    hash_suffix = hashlib.sha1(
        now.strftime(f"{DATETIME_FORMAT}%S").encode()
    ).hexdigest()[:HASH_LENGTH]

    run_copy_dir = str(
        Path(cry_copy_dir) / f"run_{now_str}_{hash_suffix}"
    )

    ignore_func = None
    if exclude_from_copy is not None:
        ignore_func = shutil.ignore_patterns(*exclude_from_copy)
    shutil.copytree(
        src=work_dir,
        dst=run_copy_dir,
        ignore=ignore_func
    )

    return run_copy_dir
```

pkuderov replied:
removed those funcs, moved base expand/sanitize funcs to the `validators.py`, reworked their application via pydantic `AfterValidator` (see `config.py`)

mrsndmn commented on cryri/utils.py:
```diff
@@ -1,64 +1,131 @@
+
+    # drop non-dir last part => dir path
+    if not p.is_dir():
+        p = p.parent
```
This silent workaround might lead to unexpected behavior. Consider throwing an exception here instead. 

Are there any cases where this silent approach would actually be preferable? If so, could we add a comment explaining why?

pkuderov replied:
Good point! I was working within the context of the previous behavior, where .parent was always used — initially, I thought the intention was to handle cases where a file path is provided. But that probably doesn’t make much sense and is unlikely to be useful, so it’s safe to remove that behavior and raise an exception instead.

mrsndmn replied:
I have never used this feature. Removing `.parent` would be a breaking change. However it would be ok for me)

@Tviskaron what do you think about it?

pkuderov replied:
added assertion

mrsndmn commented on tests/test_main.py:
```diff
@@ -55,18 +88,72 @@ def test_create_job_description_with_team(basic_config):
 
-def test_expand_vars_user(basic_config):
-    from os import environ
+def test_expand_vars_and_user(expandable_struct, caplog):
```
Thanks for adding this test!

For future reference, it’s usually better to split large test cases into smaller, atomic tests. Smaller tests are easier to maintain and debug.

mrsndmn reviewed (changes requested):
The workaround for `sanitize_dir_path` was the most frustrating part —  it should be fixed. Everything else looks great though! 👍

pkuderov committed:
make expand vars/user and sanitize dir path funcs to be pydantic fields validators; add env/filepath mocks; split/extend tests for validators

pkuderov commented:
Added env and path-resolving mocks w/ their respective tests to track their correctness. Extracted validators-related tests to corresponding file, added new file-related tests.

pkuderov committed:
remove assert, added during debug

mrsndmn commented:
Merged! @pkuderov good job!
